# OpenAI embeddings for recommendations
## 0. Imports

In [1]:
import pandas as pd
import json
from main.openai.openai_utils.openai_utils import set_embeddings_from_df

In [2]:
WORKING_DIR = "../../../beauty"

In [3]:
product_lookup = pd.read_csv(f"{WORKING_DIR}/products_lookup.csv", usecols=["global_product_id", "name"])
product_lookup.head(10)

global_product_id                                               name
0               1504  WAWO 15 Color Professionl Makeup Eyeshadow Cam...
1                564                  Xtreme Brite Brightening Gel 1oz.
2               9963  Prada Candy By Prada Eau De Parfum Spray 1.7 O...
3               9839  Versace Bright Crystal Eau de Toilette Spray f...
4               4132                            Stella McCartney Stella
5                438  Avalon Biotin B-Complex Thickening Conditioner...
6               9381  Better Living Classic Two Chamber Dispenser, W...
7               3618              Better Living The Ulti-Mate Dispenser
8              11467  Crabtree  and  Evelyn - Gardener's Ultra-Moist...
9               6073  Crabtree  and  Evelyn 2792 Gardeners Hand Ther...

In [4]:
product_lookup_empty_names = product_lookup[~product_lookup["name"].notnull()]
product_lookup = product_lookup[product_lookup["name"].notnull()]

print(f"Number of empty names: {len(product_lookup_empty_names)}")
product_lookup_empty_names.head()

Number of empty names: 7


global_product_id name
1461               1302  NaN
3578               1327  NaN
4206               5210  NaN
4381              12017  NaN
4398                122  NaN

In [5]:
step_size = 1000
batch_product_lookup = product_lookup.iloc[0:step_size]
cur_step = step_size
processed_batches = []
num_batches_processed = 0
while not(batch_product_lookup.empty):
    print(f"Currently at {cur_step} of {len(product_lookup)}")
    set_embeddings_from_df(batch_product_lookup)
    processed_batches.append(batch_product_lookup)

    batch_product_lookup = product_lookup.iloc[cur_step:cur_step + step_size]
    cur_step += step_size

new_product_lookup = pd.concat(processed_batches, copy=False)

new_product_lookup.head(10)

Currently at 1000 of 12094
Currently at 2000 of 12094
Currently at 3000 of 12094
Currently at 4000 of 12094
Currently at 5000 of 12094
Currently at 6000 of 12094
Currently at 7000 of 12094
Currently at 8000 of 12094
Currently at 9000 of 12094
Currently at 10000 of 12094
Currently at 11000 of 12094
Currently at 12000 of 12094
Currently at 13000 of 12094


/Users/anonymous/Documents/GitHub/LLM-sequential-recommendations/main/openai/openai_utils/openai_utils.py:96: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["ada_embedding"] = embeddings
/Users/anonymous/Documents/GitHub/LLM-sequential-recommendations/main/openai/openai_utils/openai_utils.py:96: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["ada_embedding"] = embeddings
/Users/anonymous/Documents/GitHub/LLM-sequential-recommendations/main/openai/openai_utils/openai_utils.py:96: SettingWithCopyWarning: 

global_product_id                                               name  \
0               1504  WAWO 15 Color Professionl Makeup Eyeshadow Cam...   
1                564                  Xtreme Brite Brightening Gel 1oz.   
2               9963  Prada Candy By Prada Eau De Parfum Spray 1.7 O...   
3               9839  Versace Bright Crystal Eau de Toilette Spray f...   
4               4132                            Stella McCartney Stella   
5                438  Avalon Biotin B-Complex Thickening Conditioner...   
6               9381  Better Living Classic Two Chamber Dispenser, W...   
7               3618              Better Living The Ulti-Mate Dispenser   
8              11467  Crabtree  and  Evelyn - Gardener's Ultra-Moist...   
9               6073  Crabtree  and  Evelyn 2792 Gardeners Hand Ther...   

                                       ada_embedding  
0  [-0.008468648418784142, 0.014345130883157253, ...  
1  [0.019681310281157494, 0.009377948939800262, -...  
2  [-0.00300808809697628, -0.007103437092155218, ...  
3  [0.0053097945638000965, 0.0017624408937990665,...  
4  [-0.006986561696976423, -0.0015255995094776154...  
5  [-0.011060410179197788, -0.017783403396606445,...  
6  [-0.004991547204554081, 0.019236043095588684, ...  
7  [-0.007757279556244612, 0.014554604887962341, ...  
8  [-0.010174826718866825, -0.0013344729086384177...  
9  [0.0050934250466525555, -0.00951578002423048, ...

In [6]:
new_product_lookup["ada_embedding"] = new_product_lookup["ada_embedding"].apply(lambda x : json.loads(x) if isinstance(x, str) else x)
new_product_lookup.head(10)

global_product_id                                               name  \
0               1504  WAWO 15 Color Professionl Makeup Eyeshadow Cam...   
1                564                  Xtreme Brite Brightening Gel 1oz.   
2               9963  Prada Candy By Prada Eau De Parfum Spray 1.7 O...   
3               9839  Versace Bright Crystal Eau de Toilette Spray f...   
4               4132                            Stella McCartney Stella   
5                438  Avalon Biotin B-Complex Thickening Conditioner...   
6               9381  Better Living Classic Two Chamber Dispenser, W...   
7               3618              Better Living The Ulti-Mate Dispenser   
8              11467  Crabtree  and  Evelyn - Gardener's Ultra-Moist...   
9               6073  Crabtree  and  Evelyn 2792 Gardeners Hand Ther...   

                                       ada_embedding  
0  [-0.008468648418784142, 0.014345130883157253, ...  
1  [0.019681310281157494, 0.009377948939800262, -...  
2  [-0.00300808809697628, -0.007103437092155218, ...  
3  [0.0053097945638000965, 0.0017624408937990665,...  
4  [-0.006986561696976423, -0.0015255995094776154...  
5  [-0.011060410179197788, -0.017783403396606445,...  
6  [-0.004991547204554081, 0.019236043095588684, ...  
7  [-0.007757279556244612, 0.014554604887962341, ...  
8  [-0.010174826718866825, -0.0013344729086384177...  
9  [0.0050934250466525555, -0.00951578002423048, ...

In [7]:
emb_dim = len(new_product_lookup.iloc[0]["ada_embedding"])
product_lookup_empty_names["ada_embedding"] = [[0.0 for _ in range(emb_dim)]] * len(product_lookup_empty_names)
product_lookup_empty_names["name"] = "Unknown item"
product_lookup_empty_names.head(10)

global_product_id          name  \
1461                1302  Unknown item   
3578                1327  Unknown item   
4206                5210  Unknown item   
4381               12017  Unknown item   
4398                 122  Unknown item   
10771               9757  Unknown item   
10990               5359  Unknown item   

                                           ada_embedding  
1461   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
3578   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
4206   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
4381   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
4398   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
10771  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
10990  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...

In [8]:
new_product_lookup.head(10)

global_product_id                                               name  \
0               1504  WAWO 15 Color Professionl Makeup Eyeshadow Cam...   
1                564                  Xtreme Brite Brightening Gel 1oz.   
2               9963  Prada Candy By Prada Eau De Parfum Spray 1.7 O...   
3               9839  Versace Bright Crystal Eau de Toilette Spray f...   
4               4132                            Stella McCartney Stella   
5                438  Avalon Biotin B-Complex Thickening Conditioner...   
6               9381  Better Living Classic Two Chamber Dispenser, W...   
7               3618              Better Living The Ulti-Mate Dispenser   
8              11467  Crabtree  and  Evelyn - Gardener's Ultra-Moist...   
9               6073  Crabtree  and  Evelyn 2792 Gardeners Hand Ther...   

                                       ada_embedding  
0  [-0.008468648418784142, 0.014345130883157253, ...  
1  [0.019681310281157494, 0.009377948939800262, -...  
2  [-0.00300808809697628, -0.007103437092155218, ...  
3  [0.0053097945638000965, 0.0017624408937990665,...  
4  [-0.006986561696976423, -0.0015255995094776154...  
5  [-0.011060410179197788, -0.017783403396606445,...  
6  [-0.004991547204554081, 0.019236043095588684, ...  
7  [-0.007757279556244612, 0.014554604887962341, ...  
8  [-0.010174826718866825, -0.0013344729086384177...  
9  [0.0050934250466525555, -0.00951578002423048, ...

In [9]:
new_product_lookup = pd.concat([new_product_lookup, product_lookup_empty_names], axis=0)

In [10]:
new_product_lookup.memory_usage(deep=True)

Index                    96808
global_product_id        96808
name                   1504157
ada_embedding        153924720
dtype: int64

In [11]:
new_product_lookup.to_csv(f"{WORKING_DIR}/product_embeddings_openai.csv.gzip", compression="gzip")

## Sanity check

In [12]:
new_product_lookup["length"] = new_product_lookup["ada_embedding"].apply(len)
new_product_lookup["type"] = new_product_lookup["ada_embedding"].apply(lambda x : type(x))

print(new_product_lookup["length"].unique())
print(new_product_lookup["type"].unique())


[1536]
[<class 'list'>]


Note that we just want "[1536]" and "[<class 'list'>]" as output here. Otherwise, the notebooks using this data will fail. 